src/Processors/Formats/Impl/Parquet/ParquetBloomFilterCondition.cpp

```diff
@@ -0,0 +1,487 @@
+
+            rpn_stack.emplace_back(maybe_true, true);
+
+            if (element.function == Function::FUNCTION_NOT_EQUALS)
+                rpn_stack.back() = !rpn_stack.back();
+        }
+        else if (element.function == Function::FUNCTION_IN
+                 || element.function == Function::FUNCTION_NOT_IN)
+        {
+            bool maybe_true = true;
+            for (auto column_index = 0u; column_index < element.columns.size(); column_index++)
+            {
+                // in case bloom filter is not present for this row group
+                // https://github.com/ClickHouse/ClickHouse/pull/62966#discussion_r1722361237
+                if (!column_index_to_column_bf.contains(element.key_columns[column_index]))
+                {
+                    rpn_stack.emplace_back(true, true);
+                    continue;
+                }
+
+                bool column_maybe_contains = maybeTrueOnBloomFilter(
+                    element.columns[column_index],
+                    column_index_to_column_bf.at(element.key_columns[column_index]),
+                    false);
+
+                if (!column_maybe_contains)
+                {
+                    maybe_true = false;
+                    break;
+                }
+            }
+
+            rpn_stack.emplace_back(maybe_true, true);
+            if (element.function == Function::FUNCTION_NOT_IN)
+                rpn_stack.back() = !rpn_stack.back();
+        }
+        else if (element.function == Function::FUNCTION_NOT)
+        {
+            rpn_stack.back() = !rpn_stack.back();
+        }
+        else if (element.function == Function::FUNCTION_OR)
+        {
+            auto arg1 = rpn_stack.back();
+            rpn_stack.pop_back();
+            auto arg2 = rpn_stack.back();
+            rpn_stack.back() = arg1 | arg2;
+        }
+        else if (element.function == Function::FUNCTION_AND)
+        {
+            auto arg1 = rpn_stack.back();
+            rpn_stack.pop_back();
+            auto arg2 = rpn_stack.back();
+            rpn_stack.back() = arg1 & arg2;
+        }
+        else if (element.function == Function::ALWAYS_TRUE)
+        {
+            rpn_stack.emplace_back(true, false);
+        }
+        else
+        {
+            rpn_stack.emplace_back(true, true);
+        }
+    }
+
+    if (rpn_stack.size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected stack size in KeyCondition::mayBeTrueOnRowGroup");
+
+    return rpn_stack[0].can_be_true;
+}
+
+std::unordered_set<std::size_t> ParquetBloomFilterCondition::getFilteringColumnKeys() const
+{
+    std::unordered_set<std::size_t> column_keys;
+
+    for (const auto & element : condition)
+    {
+        for (const auto index : element.key_columns)
+        {
+            column_keys.insert(index);
+        }
+    }
+
+    return column_keys;
+}
+
+std::vector<ParquetBloomFilterCondition::ConditionElement> keyConditionRPNToParquetBloomFilterCondition(
```

Comment:
@al13n321 I would love to get your opinion / help with this method. I am somewhat confused with the data types here.

I have modified the code so I call the proper parquet hasher API instead of calling the `parquet::ByteArray` one. Therefore, I need to be able to tell the compiler which overload to use depending on the clickhouse column type.

For simple queries like `select ... where id = 1` is simple. If `id` column is of type `uint32_t`, I do `actual_type->createColumn(); column->insert(rpn_element.range.left)` and `hashInt<uint32_t>`,  then something like `assert_cast<ColumnVector<uint32_t>>` and you know the drill. 

It gets a bit more tricky when the query is like this: `select ... where toIPv4(ip_addr) = toIPv4('192.168.0.1')`, considering `ip_addr` is of type `uint32_t`, there is a type mismatch. so: `actual_type->createColumn()` returns a column of type `uint32_t`. In order to populate this column to call the right hashing overload, I need to first convert the `rpn_element.range.left`, that contains a IPv4 value, to uint32_t. I have tried using `tryConvertFieldToType`, but for some reason it is not allowed. At least it does not crash and filters it correctly (not using BF, ofc)

It is even trickier when ipv4 is in a tuple. E.g, `select ... where (toIPv4(ip_addr)) in (toIPv4('192.168.0.1'))`. There is no range in the `rpn_element`. What can be found is a set of columns, which include a single IPv4 column. I could do something like `canBeSafelyCasted()` and then `castColumn`, but it is starting to sound like a rabbit hole.

Do you have a suggestion?

Reply:
Actually, I think it does not make sense to support `select ... where toIPv4(ip_addr) = toIPv4('192.168.0.1')`. Filtering is based on a transformation of the raw column, there is no raw value to "look for".

The commit I just pushed basically relies on `tryConvertFieldToType`.